cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -242,5 +240,15 @@ Result<ScanTaskIterator> CsvFileFormat::ScanFile(
   return MakeVectorIterator<std::shared_ptr<ScanTask>>({std::move(task)});
 }
 
+Result<RecordBatchGenerator> CsvFileFormat::ScanBatchesAsync(
+    const std::shared_ptr<ScanOptions>& scan_options,
+    const std::shared_ptr<FileFragment>& file) const {
+  auto this_ = checked_pointer_cast<const CsvFileFormat>(shared_from_this());
+  auto source = file->source();
+  auto reader_fut =
+      OpenReaderAsync(source, *this, scan_options, internal::GetCpuThreadPool());
+  return GeneratorFromReader(std::move(reader_fut));
```

Comment:
Should we go ahead and add readahead here?

Reply:
The CSV reader does not support parallel readahead yet.  Serial readahead might be redundant since the background generator is pretty much already doing that.  I suppose it would give us a bit of pipeline parallelism (allowing us to filter & project while we parse & decode) but at the cost breaking up some of the processing locality (e.g. right now we always parse X, decode X, filter X, project X).  I'll run some experiments and see if it has any noticeable effect in either direction.

Reply:
Adding serial readahead here did not have any effect on performance.